tutorials/phasorpy_lfd_workshop.py

```diff
@@ -220,17 +221,13 @@
 # In addition, small amounts of room light will appear towards the lower
 # left-hand corner of the phasor (room light is uncorrelated, so it has
 # zero modulation depth). This can be eliminated by setting a threshold.
-#
-# .. todo::
-#
-#     Use a ``phasor_threshold`` function?
-
-threshold = (mean > 20) & (real1 > 0) & (imag1 > 0) & (real2 > 0) & (imag2 > 0)
-real1 = real1[threshold]
-imag1 = imag1[threshold]
-real2 = real2[threshold]
-imag2 = imag2[threshold]
 
+_, real1, imag1 = phasor_threshold(
+    mean, real1, imag1, 20, real_min=0, imag_min=0
+)
+_, real2, imag2 = phasor_threshold(
+    mean, real2, imag2, 20, real_min=0, imag_min=0
+)
```

Comment:
This is OK, although again technically not the same. How should filtering for example `mean <= 0.0` instead of `mean < 0.0` be handled? Maybe `mean < 1e-9` is good enough?

Also, is there a way to put this in a single statement with broadcasting? Just wondering.

Reply:
> This is OK, although again technically not the same. How should filtering for example `mean <= 0.0` instead of `mean < 0.0` be handled?

This is a very good question also. Before, I had implemented it so that the lower bound is included, but I just realized that the new implementation doesn't consider the bounds. I think it's better to include the bounds (similar to what happens with the masking of different shapes you implemented, but in that case it is only included the lower bound). I think it also makes sense considering the name of the parameters, meaning that `_min` is the minimum value that is not thresholded. I don't know how other libraries that apply threshold in imaging handle this, but I will research and maybe we can discuss the better approach further. 


> Also, is there a way to put this in a single statement with broadcasting? Just wondering.

I left it like this since to try to keep the same structure as before, but I will look into it.

Reply:
> I left it like this since to try to keep the same structure as before, but I will look into it.

I am just curious since multiple harmonics are returned by `phasor_from_signal` in one array.

Re bounds: There are certainly arguments for both <=, >= vs <, >. My reference was to keep it similar to `numpy.clip`, which looks a lot like our function. Also, we consider the phasor coordinates 0.0 and 1.0 valid, while > 1.0 and < 0.0 (for calibrated lifetimes) are not, and the function is probably more often used to remove "invalid" values. If needed, we could add a switch like `open_interval: bool = ...` and have two sets of Cython functions?

Reply:
> I think it's better to include the bounds (similar to what happens with the masking of different shapes you implemented, but in that case it is only included the lower bound

Those functions seem consistent with what is implemented here. Values outside the closed interval [min..max] are rejected (False or NaN). No? With the exception of the range function. The argument for range to be different from interval is that (1) Python range excludes the upper value and (2) several ranges may be stacked together. Although we may want to review if that makes sense for floating point.

Reply:
Those are all good points you raised Christoph. I think every approach has it's advantages and disadvantages, but maybe the best is to add the `open_interval` to make as flexible as possible without adding that much complexity?

> we consider the phasor coordinates 0.0 and 1.0 valid, while > 1.0 and < 0.0 (for calibrated lifetimes) are not, and the function is probably more often used to remove "invalid" values.

Also considering this, it makes a lot of sense to set by default to include the bounds, as not to exclude the 1.0 and 0.0. Do you think this can be the best approach?

Reply:
Sorry @cgohlke I misunderstood the approach I think. You are right, the current implementation considers the closed interval (the default condition), but maybe it would be nice to add the option to make it open?

Reply:
Maybe I am missing something, but the current implementation does not exclude min or max. It returns NaN only if a value `< min` or `> max`. No?

Reply:
Exactly! I was wrong and I thought it was the other way. But it actually does that currently. Should we keep like this or add the `open_interval`?

Reply:
> Should we keep like this or add the open_interval?

I leave that up to you. It's straightforward to implement but needs to be tested and documented too.